### ISSUE TITLE
Handling of Unicode suggestions

### ISSUE THREAD

**Issue by dmoonfire** [atom/node-spellchecker#77]
I'm looking at a couple problems with Atom's `spell-check` module and I think I've traced a problem down to this library. In specific, passing in words to `GetCorrectionsForMisspelling` doesn't seem to handle encoding correctly.

In this case, I'm using a German word `möchten` which is spelled correctly. If I run the test via the command line (Ubuntu with a `de-DE` dictionary installed), I get this:

```shell
$ hunspell -d /usr/share/hunspell/de_DE
Hunspell 1.4.0
möchten
*

möchte 
+ möchten

$
```

This is pretty much the output I'm expecting to see. If I do the same with a short JS script, however:

```javascript
// Create the spelling wrapper and add the German dictionary to it.
var spell = require("./lib/spellchecker");
spell.setDictionary("de-DE", '/usr/share/hunspell');

// Test some function calls.
console.log("isMisspelled:", spell.isMisspelled("möchten"));
console.log("checkSpelling:", spell.checkSpelling("möchten"));
console.log("getCorrectionsForMisspelling:", spell.getCorrectionsForMisspelling("möchten"));
```
I get the following:

```shell
$ node test.js
isMisspelled: true
checkSpelling: [ ]
getCorrectionsForMisspelling: [ 'm�chten' ]
$
```

From the CLI version, this isn't a misspelled word, it isn't marked as an error though (`checkSpelling`), but the suggestions don't come through. I noticed that `checkSpelling` has a slightly more complex signature for converting from UTF-16 to UTF-8 but the others don't.

`isMispelled` only has a boolean output, which kind of suggests that the Unicode string going in is not being translated in hunspell's expected format. Since "möchten" is correct, I suspect that the same problem exists in `getCorrectionsForMisspelling` in addition to the output converted might need encoding.

I looked at the resulting hex output from the `getCorrectionsForMisspelling` and got `<Buffer 6d ef bf bd 63 68 74 65 6e>`. The `ef bf bd` should have been the umlaut character, however it is being returned as a [placeholder character](https://apps.timwhitlock.info/unicode/inspect?s=%EF%BF%BD) in UTF-8 encoding. I think this suggests that we might be dealing with a different encoding there.

Sadly, I couldn't figure it out. I spent a few hours messing with the gyp and C++ but haven't figured out the specifics to get it working. I tried converting `getCorrectionsForMisspelling` to use the UTF-16 format but I don't think I did it correctly since I'm just monkey-bashing code to see if to works.

Any help would be appreciated.

**Comment by scottnonnenberg** [comment]
It's interesting to see how much work is done in the hunspell `CheckSpelling()` wrapper vs. the other method wrappers: https://github.com/atom/node-spellchecker/blob/edf4678e2853dae42630f2b6024c1311c461b0fc/src/spellchecker_hunspell.cc#L57

`CheckSpelling()` takes its parameters as UTF16, where the others just take `const std::string& word`. And based on this, all strings in the Javascript environment start out as UTF16. https://kev.inburke.com/kevin/node-js-string-encoding/

So I suppose the solution would be to treat all incoming strings in `spellechecker_hunspell.css` as UTF16, then convert them to UTF8 before sending them to hunspell?

**Comment by dmoonfire** [comment]
I'm looking at this again because there are a number of issues being opened up on atom/spell-check. I'm not sure if the UTF16 to UTF8 is the entire answer. I've added a few new tests, mostly using the German dictionary and trying to get the tests that break to help isolate the problem. I think I have the tests correct, a bit of validation wouldn't hurt there. :)

*I noticed that the tests use `de_DE` but my Linux machine needed `de_DE_frami` to work properly. Not sure if that is right.*

I ended up focusing on a single test using `jasmine-focused` and `fit` inside the test to test for "Kein Kine möchten möchte" (Kein and möchten are correct, Kine and möchte are not). Running those tests, I added a number of `std::count` to `spellchecker_hunspell.cpp` to try identifying the problem.  (All my debugging is prefixed with `DREM `).

So far what I've noticed is that the loop doesn't see the UTF16 words. It gets an "unknown" state and blows up. However, if I add `setlocale(LC_ALL, "en_US.utf8");`, then it picks them up however it doesn't correctly identify the words.

At this, I'm still not sure how to fix this still.

My current WIP is at https://github.com/dmoonfire/node-spellchecker/tree/unicode-checking.

**Comment by rmarianni** [comment]
@dmoonfire I might have found something interesting...

Curiously enough, this also happens when I try to [use the Hunspell library with GCC](https://github.com/hunspell/hunspell#using-hunspell-library-with-gcc).

I'm working on an Electron application, and one of the requirements is spell checking. I could not build the application using [electron-spellchecker](https://github.com/electron-userland/electron-spellchecker) for some reasons, so I've decided to look for another library. I've found [nodehun](https://github.com/Wulf/nodehun). Nodehun has its own [native module implementation](https://github.com/Wulf/nodehun/blob/master/src/post0.12.0/nodehun.cpp), but it only took a few unit tests to find out that I have the exact same errors you have, when I try to verify accented Brazilian Portuguese words.

I've decided to check if the same errors happen with the C++ library, and they do (my `.cxx` file's encoding was UTF-8, and I've tested the library with both accented chars and `\u` codes in the literals).

Then I remembered I should check the encoding of my local hunspell `.aff` and `.dic` files (at `/usr/share/hunspell`), and I was surprised to see that both are encoded in ISO-8859 (-1, probably). 

So I've figured out that I could test two other things:

1. I could re-encode the `.cxx` file to ISO-8859-1 and test my program again. Unsurprisingly, all results were correct after I've done this (except for the output, since my terminal is naturally configured to use the UTF-8 charset).

2. Better than that, I could re-encode the `.aff` and `.dic` files to UTF-8, and leave my `.cxx` file in UTF-8. This resulted in multiple console errors of the form `error: line 1176: multiple definitions of an affix flag`. A quick search took me to a very interesting [bug report](https://sourceforge.net/p/hunspell/bugs/138/) at hunspell... So I've followed [Caolan Mcnamara's advice](https://sourceforge.net/p/hunspell/bugs/138/#7829) and added the `FLAG=UTF-8` in my UTF-8-encoded `.aff` file, and guess what... Everything worked as expected - both my C++ program **and** my Electron/Node.js program...

It seems all I have to do now is to distribute this "fixed" UTF-8-encoded `.aff` file with my UTF-8-encoded `.dic` file, and I'll have correct spell checking in my Electron application... :)

By the way, if this is actually fixes things for other people too, maybe there is another thing to consider: [these UTF-8 hunspell dictionaries](https://github.com/titoBouzout/Dictionaries) do not have the FLAG=UTF-8 in their `.aff` files; well, the Brazilian Portuguese one doesn't. Considering that hunspell bug report, though, I think maybe they should add it, and the [encoding instructions](https://github.com/titoBouzout/Dictionaries#adding-a-new-language) might be updated too.

I don't know if this information is useful for you, but I hope this helps somehow.

**Comment by dmoonfire** [comment]
@s-m-e: Yeah, I realized I was using the wrong word. My latest efforts in https://github.com/atom/node-spellchecker/pull/95 use a different word. I had a problem with the `setlocale` call inside the Hunspell code in relationship to the dictionaries. If both the file and the locale are UTF8, it works just fine but I'm not sure how to fix it properly that doesn't require reencoding the dictionaries.

Also, those fixes don't work when using the built-in spell checker for Windows 10 or Mac OS. I'm plugging at it, but C isn't my native language anymore and it's taking me a while. I apologize.

**Comment by s-m-e** [comment]
@dmoonfire What's the current status with respect to this bug?

**Comment by lee-dohm** [comment]
@maxbrunsfeld do you have any bandwidth to help out with this? I have a sneaking suspicion that this might be an issue around text-buffer?
